Add support for largeBlob extension for local authenticator

g-davidson opened the pull request:
#### a38ad3411dcf6207dd198956e375a5d7302323f9
<pre>
Add support for largeBlob extension for local authenticator
<a href="https://bugs.webkit.org/show_bug.cgi?id=252789">https://bugs.webkit.org/show_bug.cgi?id=252789</a>
rdar://105237759

Reviewed by J Pascoe.

This patch adds support for the largeBlob extension to the local authenticator. This
extension allows storing an arbitrary blob of data alongside a passkey, which can be read
or written during assertions.

* Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
(JSC::ArrayBuffer::create):
* Source/JavaScriptCore/runtime/ArrayBuffer.h:
Moved Vector helpers from WebAuthnUtils to be first class methods.

* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.cpp:
(WebCore::AuthenticationExtensionsClientInputs::fromCBOR):
(WebCore::AuthenticationExtensionsClientInputs::toCBOR const):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.h:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.idl:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.cpp:
(WebCore::AuthenticationExtensionsClientOutputs::fromCBOR):
(WebCore::AuthenticationExtensionsClientOutputs::toCBOR const):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.h:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.idl:
Basic IDL and encoding support.

* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.h:
(WebCore::AuthenticatorAssertionResponse::largeBlob const):
(WebCore::AuthenticatorAssertionResponse::setLargeBlob):
Hold on to the blob when we load credentials.

* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
(WebCore::coseKeyForAttestationObject):
(WebCore::AuthenticatorAttestationResponse::getAuthenticatorData const):
Switch away from the old helpers.

* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::create):
Remove an outdated comment and pass along the new extension.

* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
(WebCore::convertArrayBufferToVector): Deleted.
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.h:

* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
(fido::encodeMakeCredentialRequestAsCBOR):
(fido::encodeGetAssertionRequestAsCBOR):
(fido::encodeMakeCredenitalRequestAsCBOR): Deleted.
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.h:
Add support for serializing the extension during get and create. This also checks to make
sure that the authenticator supports the extension before attempting to serialize it.
Comes with tests!

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
Support for serializing the new extension between the web process and the UI process. The
also requires the ability to serialize ArrayBuffer to match the IDL.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(toVector):
(+[_WKWebAuthenticationPanel encodeMakeCredentialCommandWithClientDataJSON:options:userVerificationAvailability:]):
(+[_WKWebAuthenticationPanel encodeMakeCredentialCommandWithClientDataJSON:options:userVerificationAvailability:authenticatorSupportedExtensions:]):
(+[_WKWebAuthenticationPanel encodeGetAssertionCommandWithClientDataJSON:options:userVerificationAvailability:]):
(+[_WKWebAuthenticationPanel encodeGetAssertionCommandWithClientDataJSON:options:userVerificationAvailability:authenticatorSupportedExtensions:]):
(+[_WKWebAuthenticationPanel encodeMakeCredentialCommandWithClientDataHash:options:userVerificationAvailability:]):
(+[_WKWebAuthenticationPanel encodeMakeCredentialCommandWithClientDataHash:options:userVerificationAvailability:authenticatorSupportedExtensions:]):
(+[_WKWebAuthenticationPanel encodeGetAssertionCommandWithClientDataHash:options:userVerificationAvailability:]):
(+[_WKWebAuthenticationPanel encodeGetAssertionCommandWithClientDataHash:options:userVerificationAvailability:authenticatorSupportedExtensions:]):
Trivial new versions of all of these methods that accept an array of extensions supported
by the authenticator.

* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::AuthenticatorManager::respondReceived):
When the extension fails, the spec says to return a NotSupportedError, which is fatal.
Update this method to treate it as such.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(alternateBlobIfNecessary):

(WebKit::LocalAuthenticatorInternal::getExistingCredentials):
When loading credentials, also fetch the new blob.

(WebKit::LocalAuthenticator::processLargeBlobExtension):
This extension is now always supported.
If we&apos;re doing a read, populate the output with the loaded blob.
If we&apos;re doing a write, fetch the existing credential data structure, create a copy with
the new blob populated, and write it back to the keychain.

(WebKit::LocalAuthenticator::processClientExtensions):
The largeBlob extension can fail, causing it to return a NotSupportedError, which is
fatal. This method now returns an exception if something went wrong when processing any
extension, or nullopt on success.

(WebKit::LocalAuthenticator::continueMakeCredentialAfterUserVerification):
(WebKit::LocalAuthenticator::continueMakeCredentialAfterAttested):
(WebKit::LocalAuthenticator::getAssertion):
(WebKit::LocalAuthenticator::continueGetAssertionAfterUserVerification):
Update these to handle exceptions during extension processing.

* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::makeCredential):
(WebKit::CtapAuthenticator::getAssertion):
* Tools/TestWebKitAPI/Tests/WebCore/CtapRequestTest.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebCore/CtapResponseTest.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebCore/FidoTestData.h:

Canonical link: <a href="https://commits.webkit.org/260958@main">https://commits.webkit.org/260958@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9d5ae2d102ceb62a5a6c61339b8ea114ff9d6ea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109826 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42555 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1288 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118927 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113776 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20406 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10135 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102095 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115574 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15192 "Found 1 new test failure: http/wpt/webauthn/public-key-credential-create-failure-local-silent.https.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98420 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43419 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97155 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30077 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85220 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/98676 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11658 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31419 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/99568 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/9691 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12302 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8358 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31139 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17672 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51028 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/107586 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7599 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14064 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26551 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
```diff
@@ -212,6 +212,11 @@ Ref<ArrayBuffer> ArrayBuffer::create(ArrayBufferContents&& contents)
     return adoptRef(*new ArrayBuffer(WTFMove(contents)));
 }
 
+Ref<ArrayBuffer> ArrayBuffer::create(const Vector<uint8_t> vector)
```
I think this will do an unnecessary copy. `const Vector<uint8_t>&`

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.cpp:
```diff
@@ -50,6 +50,28 @@ std::optional<AuthenticationExtensionsClientInputs> AuthenticationExtensionsClie
+
+        largeBlobIt = largeBlobMap.find(cbor::CBORValue("write"));
+        if (largeBlobIt != decodedMap.end() && largeBlobIt->second.isByteString()) {
+            RefPtr<ArrayBuffer> write = ArrayBuffer::create(largeBlobIt->second.getByteString().data(), largeBlobIt->second.getByteString().size());
```
I think you probably want to use the helper you defined in ArrayBuffer here. ByteString is a CBOR BinaryValue, which is defined as `using BinaryValue = Vector<uint8_t>;`

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.cpp:
```diff
@@ -53,6 +53,28 @@ std::optional<AuthenticationExtensionsClientOutputs> AuthenticationExtensionsCli
+
+        largeBlobIt = largeBlobMap.find(cbor::CBORValue("blob"));
+        if (largeBlobIt != decodedMap.end() && largeBlobIt->second.isByteString()) {
+            RefPtr<ArrayBuffer> blob = ArrayBuffer::create(largeBlobIt->second.getByteString().data(), largeBlobIt->second.getByteString().size());
```
ditto

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
```diff
@@ -38,7 +38,7 @@ namespace WebCore {
 static std::optional<cbor::CBORValue> coseKeyForAttestationObject(Ref<ArrayBuffer> attObj)
 {
-    auto decodedResponse = cbor::CBORReader::read(convertArrayBufferToVector(attObj.ptr()));
+    auto decodedResponse = cbor::CBORReader::read(WTFMove(attObj->vector()));
```
I don't think you need the move here.

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
```diff
@@ -96,7 +96,7 @@ AuthenticatorResponseData AuthenticatorAttestationResponse::data() const
 RefPtr<ArrayBuffer> AuthenticatorAttestationResponse::getAuthenticatorData() const
 {
-    auto decodedResponse = cbor::CBORReader::read(convertArrayBufferToVector(m_attestationObject.ptr()));
+    auto decodedResponse = cbor::CBORReader::read(WTFMove(m_attestationObject->vector()));
```
I don't think you need the move here.

pascoej commented on Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
```diff
@@ -86,7 +86,7 @@ static CBORValue convertDescriptorToCBOR(const PublicKeyCredentialDescriptor& de
 }
 
-Vector<uint8_t> encodeMakeCredenitalRequestAsCBOR(const Vector<uint8_t>& hash, const PublicKeyCredentialCreationOptions& options, UVAvailability uvCapability, AuthenticatorSupportedOptions::ResidentKeyAvailability residentKeyAvailability, std::optional<PinParameters> pin)
+Vector<uint8_t> encodeMakeCredentialRequestAsCBOR(const Vector<uint8_t>& hash, const PublicKeyCredentialCreationOptions& options, UVAvailability uvCapability, AuthenticatorSupportedOptions::ResidentKeyAvailability residentKeyAvailability, const Vector<String>& authenticatorSupportedExtensions, std::optional<PinParameters> pin)
```
🎉

pascoej commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -3923,3 +3937,8 @@ header: <WebCore/SecurityPolicyViolationEvent.h>
     unsigned long lineNumber();
     unsigned long columnNumber();
 };
+
```
Is this needed?

g-davidson replied:
It's used by `WebCore::AuthenticationExtensionsClientOutputs::LargeBlobOutputs` above in this file

pascoej commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.h:
```diff
@@ -145,10 +145,14 @@ WK_CLASS_AVAILABLE(macos(10.15.4), ios(13.4))
 
 + (NSData *)getClientDataJSONForAuthenticationType:(_WKWebAuthenticationType)type challenge:(NSData *)challenge origin:(NSString *)origin WK_API_AVAILABLE(macos(13.0), ios(16.0));
 + (NSData *)encodeMakeCredentialCommandWithClientDataJSON:(NSData *)clientDataJSON options:(_WKPublicKeyCredentialCreationOptions *)options userVerificationAvailability:(_WKWebAuthenticationUserVerificationAvailability)userVerificationAvailability WK_API_AVAILABLE(macos(13.0), ios(16.0));
++ (NSData *)encodeMakeCredentialCommandWithClientDataJSON:(NSData *)clientDataJSON options:(_WKPublicKeyCredentialCreationOptions *)options userVerificationAvailability:(_WKWebAuthenticationUserVerificationAvailability)userVerificationAvailability authenticatorSupportedExtensions:(nullable NSArray<NSString *> *)authenticatorSupportedExtensions WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
Is the contents of authenticatorSupportedExtensions going to be static for the platform authenticator?

g-davidson replied:
Yes, though this method is also used for hybrid where it won't be static.

cdumez commented on Source/JavaScriptCore/runtime/ArrayBuffer.h:
```diff
@@ -320,6 +321,8 @@ class ArrayBuffer final : public GCIncomingRefCounted<ArrayBuffer> {
     Expected<int64_t, GrowFailReason> grow(VM&, size_t newByteLength);
     Expected<int64_t, GrowFailReason> resize(VM&, size_t newByteLength);
 
+    Vector<uint8_t> vector() const { return { static_cast<const uint8_t*>(data()), byteLength() }; };
```
cc @Constellation to review the ArrayBuffer changes.

Constellation replied:
Let's name it `toVector()`. The name `vector()` makes me think that it is a getter returning a `m_vector` field.
Also, new data is allocated and copied. Is it expected? Maybe copying is fine. If you would like to use direct pointer and size, then you need to be extra careful since ArrayBuffer is resizable and growable. If you invoke some JS operations, then pointer can be invalidated (and byteLength() can be changed concurrently). So, maybe copying is fine here for safety.

g-davidson replied:
I hadn't considered the size changes, but given that concern I agree that making a copy sounds like a better idea.

cdumez commented on Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.h:
```diff
@@ -27,14 +27,22 @@
 
 struct AuthenticationExtensionsClientInputs {
+    struct LargeBlobInputs {
+        std::optional<String> support;
```
We don't use `std::optional<String>`. String already has a null state so String::isNull() will be true when the support member wasn't provided.

cdumez commented on Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.idl:
```diff
@@ -23,11 +23,20 @@
 
+ [
+    Conditional=WEB_AUTHN,
+] dictionary LargeBlobInputs {
```
It seems to be called `AuthenticationExtensionsLargeBlobInputs` in the spec. Why are we using a different name?

g-davidson replied:
Since this is already nested under `AuthenticationExtensionsClientInputs` the longer name seemed redundant. I'm happy to use though if that's the preferred style.

cdumez replied:
The IDL should match the spec ideally though you can use a shorter / different name on native side using [ImplementedAs] in IDL.

g-davidson replied:
Oh perfect, thanks!

cdumez commented on Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.idl:
```diff
@@ -29,10 +29,20 @@
+[
+    Conditional=WEB_AUTHN,
+    JSGenerateToJSObject,
+] dictionary LargeBlobOutputs {
```
Seems to be called AuthenticationExtensionsLargeBlobOutputs in the spec?

g-davidson replied:
Ditto

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
```diff
@@ -112,6 +112,20 @@ static inline void updateQueryForGroupIfNecessary(NSMutableDictionary *dictionar
     return adoptNS([[NSData alloc] initWithBytes:data.data() length:data.size()]);
 }
+
+static inline Vector<String> toVector(NSArray<NSString *> *strings)
```
There is already a makeVector in VectorCocoa.h

Constellation commented on Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
```diff
@@ -212,6 +212,11 @@ Ref<ArrayBuffer> ArrayBuffer::create(ArrayBufferContents&& contents)
+Ref<ArrayBuffer> ArrayBuffer::create(const Vector<uint8_t> vector)
+{
+    return ArrayBuffer::create(vector.data(), vector.size());
+}
```
I cannot find the use of this function in this patch. If it is not used, then remove it.

g-davidson replied:
It's used now!

Constellation commented on Source/JavaScriptCore/runtime/ArrayBuffer.h:
```diff
@@ -260,6 +260,7 @@ class ArrayBuffer final : public GCIncomingRefCounted<ArrayBuffer> {
     JS_EXPORT_PRIVATE static Ref<ArrayBuffer> create(ArrayBuffer&);
     JS_EXPORT_PRIVATE static Ref<ArrayBuffer> create(const void* source, size_t byteLength);
     JS_EXPORT_PRIVATE static Ref<ArrayBuffer> create(ArrayBufferContents&&);
+    JS_EXPORT_PRIVATE static Ref<ArrayBuffer> create(const Vector<uint8_t>);
```
Ditto.

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.cpp:
```diff
@@ -50,6 +50,28 @@ std::optional<AuthenticationExtensionsClientInputs> AuthenticationExtensionsClie
+            largeBlob.support = largeBlobIt->second.getString();
+
+        largeBlobIt = largeBlobMap.find(cbor::CBORValue("read"));
+        if (largeBlobIt != decodedMap.end() && largeBlobIt->second.isBool())
```
`largeBlobIt != decodedMap.end() `

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.cpp:
```diff
@@ -50,6 +50,28 @@ std::optional<AuthenticationExtensionsClientInputs> AuthenticationExtensionsClie
+            largeBlob.read = largeBlobIt->second.getBool();
+
+        largeBlobIt = largeBlobMap.find(cbor::CBORValue("write"));
+        if (largeBlobIt != decodedMap.end() && largeBlobIt->second.isByteString()) {
```
ditto

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.cpp:
```diff
@@ -53,6 +53,28 @@ std::optional<AuthenticationExtensionsClientOutputs> AuthenticationExtensionsCli
+        LargeBlobOutputs largeBlob;
+
+        auto largeBlobIt = largeBlobMap.find(cbor::CBORValue("supported"));
+        if (largeBlobIt != decodedMap.end() && largeBlobIt->second.isBool())
```
ditto

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.cpp:
```diff
@@ -53,6 +53,28 @@ std::optional<AuthenticationExtensionsClientOutputs> AuthenticationExtensionsCli
+            largeBlob.supported = largeBlobIt->second.getBool();
+
+        largeBlobIt = largeBlobMap.find(cbor::CBORValue("blob"));
+        if (largeBlobIt != decodedMap.end() && largeBlobIt->second.isByteString()) {
```
ditto

pascoej commented on Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.cpp:
```diff
@@ -53,6 +53,28 @@ std::optional<AuthenticationExtensionsClientOutputs> AuthenticationExtensionsCli
+        }
+
+        largeBlobIt = largeBlobMap.find(cbor::CBORValue("written"));
+        if (largeBlobIt != decodedMap.end() && largeBlobIt->second.isBool())
```
ditto

JonWBedard commented on metadata/contributors.json:
```diff
@@ -2813,6 +2813,12 @@
       ],
       "status" : "committer"
    },
+   {
```
This change needs to be broken off into a separate PR. Also, for this to be useful, we really need your GitHub username too, that's how we grant permissions in the WebKit project to accounts controlled by individuals, rather than just accounts.

pascoej reviewed (approved):
Looks good to me. The failure in iOS-wk2 looks like it needs re-baselining.

Commit:
Add support for largeBlob extension for local authenticator
https://bugs.webkit.org/show_bug.cgi?id=252789
rdar://105237759

Reviewed by J Pascoe.

This patch adds support for the largeBlob extension to the local authenticator. This
extension allows storing an arbitrary blob of data alongside a passkey, which can be read
or written during assertions.

* Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
(JSC::ArrayBuffer::create):
* Source/JavaScriptCore/runtime/ArrayBuffer.h:
Moved Vector helpers from WebAuthnUtils to be first class methods.

* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.cpp:
(WebCore::AuthenticationExtensionsClientInputs::fromCBOR):
(WebCore::AuthenticationExtensionsClientInputs::toCBOR const):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.h:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientInputs.idl:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.cpp:
(WebCore::AuthenticationExtensionsClientOutputs::fromCBOR):
(WebCore::AuthenticationExtensionsClientOutputs::toCBOR const):
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.h:
* Source/WebCore/Modules/webauthn/AuthenticationExtensionsClientOutputs.idl:
Basic IDL and encoding support.

* Source/WebCore/Modules/webauthn/AuthenticatorAssertionResponse.h:
(WebCore::AuthenticatorAssertionResponse::largeBlob const):
(WebCore::AuthenticatorAssertionResponse::setLargeBlob):
Hold on to the blob when we load credentials.

* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
(WebCore::coseKeyForAttestationObject):
(WebCore::AuthenticatorAttestationResponse::getAuthenticatorData const):
Switch away from the old helpers.

* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::create):
Remove an outdated comment and pass along the new extension.

* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
(WebCore::convertArrayBufferToVector): Deleted.
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.h:

* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
(fido::encodeMakeCredentialRequestAsCBOR):
(fido::encodeGetAssertionRequestAsCBOR):
(fido::encodeMakeCredenitalRequestAsCBOR): Deleted.
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.h:
Add support for serializing the extension during get and create. This also checks to make
sure that the authenticator supports the extension before attempting to serialize it.
Comes with tests!

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
Support for serializing the new extension between the web process and the UI process. The
also requires the ability to serialize ArrayBuffer to match the IDL.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(toVector):
(+[_WKWebAuthenticationPanel encodeMakeCredentialCommandWithClientDataJSON:options:userVerificationAvailability:]):
(+[_WKWebAuthenticationPanel encodeMakeCredentialCommandWithClientDataJSON:options:userVerificationAvailability:authenticatorSupportedExtensions:]):
(+[_WKWebAuthenticationPanel encodeGetAssertionCommandWithClientDataJSON:options:userVerificationAvailability:]):
(+[_WKWebAuthenticationPanel encodeGetAssertionCommandWithClientDataJSON:options:userVerificationAvailability:authenticatorSupportedExtensions:]):
(+[_WKWebAuthenticationPanel encodeMakeCredentialCommandWithClientDataHash:options:userVerificationAvailability:]):
(+[_WKWebAuthenticationPanel encodeMakeCredentialCommandWithClientDataHash:options:userVerificationAvailability:authenticatorSupportedExtensions:]):
(+[_WKWebAuthenticationPanel encodeGetAssertionCommandWithClientDataHash:options:userVerificationAvailability:]):
(+[_WKWebAuthenticationPanel encodeGetAssertionCommandWithClientDataHash:options:userVerificationAvailability:authenticatorSupportedExtensions:]):
Trivial new versions of all of these methods that accept an array of extensions supported
by the authenticator.

* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::AuthenticatorManager::respondReceived):
When the extension fails, the spec says to return a NotSupportedError, which is fatal.
Update this method to treate it as such.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(alternateBlobIfNecessary):

(WebKit::LocalAuthenticatorInternal::getExistingCredentials):
When loading credentials, also fetch the new blob.

(WebKit::LocalAuthenticator::processLargeBlobExtension):
This extension is now always supported.
If we're doing a read, populate the output with the loaded blob.
If we're doing a write, fetch the existing credential data structure, create a copy with
the new blob populated, and write it back to the keychain.

(WebKit::LocalAuthenticator::processClientExtensions):
The largeBlob extension can fail, causing it to return a NotSupportedError, which is
fatal. This method now returns an exception if something went wrong when processing any
extension, or nullopt on success.

(WebKit::LocalAuthenticator::continueMakeCredentialAfterUserVerification):
(WebKit::LocalAuthenticator::continueMakeCredentialAfterAttested):
(WebKit::LocalAuthenticator::getAssertion):
(WebKit::LocalAuthenticator::continueGetAssertionAfterUserVerification):
Update these to handle exceptions during extension processing.

* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::makeCredential):
(WebKit::CtapAuthenticator::getAssertion):
* Tools/TestWebKitAPI/Tests/WebCore/CtapRequestTest.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebCore/CtapResponseTest.cpp:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebCore/FidoTestData.h:

Canonical link: https://commits.webkit.org/260958@main
